src/network_helper_win.rs

```diff
@@ -0,0 +1,109 @@
+unsafe fn u16_ptr_to_string(ptr: *const u16) -> OsString {
+    let len = (0..).take_while(|&i| *ptr.offset(i) != 0).count();
+    let slice = std::slice::from_raw_parts(ptr, len);
+
+    OsString::from_wide(slice)
+}
+
+impl Iterator for InterfaceAddressIterator {
+    type Item = (String, MacAddr);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.adapter.is_null() {
+            return None;
+        }
+        unsafe {
+            let adapter = self.adapter;
+            // Move to the next adapter
+            self.adapter = (*adapter).Next;
+            if let Ok(interface_name) = u16_ptr_to_string((*adapter).FriendlyName).into_string() {
+                // take the first 6 bytes and return the MAC address instead
+                let [mac @ .., _, _] = (*adapter).PhysicalAddress;
+                Some((interface_name, MacAddr::from(mac)))
+            } else {
+                // Not sure whether error can occur when parsing adapter name.
+                self.next()
+            }
+        }
+    }
+}
+
+impl Drop for InterfaceAddressIterator {
+    fn drop(&mut self) {
+        unsafe {
+            libc::free(self.buf as _);
+        }
+    }
+}
+
+pub(crate) fn get_interface_address() -> Result<InterfaceAddressIterator, String> {
+    unsafe {
+        // https://learn.microsoft.com/en-us/windows/win32/api/iphlpapi/nf-iphlpapi-getadaptersaddresses#remarks
+        // A 15k buffer is recommended
+        let mut size: u32 = 15 * 1024;
+        let mut buf = null_mut();
+        let mut ret = ERROR_SUCCESS;
+
+        // https://learn.microsoft.com/en-us/windows/win32/api/iphlpapi/nf-iphlpapi-getadaptersaddresses#examples
+        // Try to retrieve adapter information up to 3 times
+        for _ in 0..3 {
+            buf = libc::malloc(size as usize) as PIP_ADAPTER_ADDRESSES;
+            if buf.is_null() {
+                // insufficient memory available
+                // https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/malloc?view=msvc-170#return-value
+                // malloc is not documented to set the last-error code
+                return Err("failed to allocate memory for IP_ADAPTER_ADDRESSES".to_string());
+            }
+
+            ret = GetAdaptersAddresses(
+                AF_UNSPEC as u32,
+                GAA_FLAG_SKIP_MULTICAST | GAA_FLAG_SKIP_ANYCAST | GAA_FLAG_SKIP_DNS_SERVER,
+                null_mut(),
+                buf,
+                &mut size,
+            );
+
+            if ret == ERROR_BUFFER_OVERFLOW {
+                // if the given memory size is too small to hold the adapter information,
+                // the SizePointer returned will point to the required size of the buffer,
+                // and we should continue.
+                libc::free(buf as _);
+                buf = null_mut();
+            } else {
+                // Otherwise, break the loop and check the return code again
+                break;
+            }
```

Comment:
```suggestion
            if ret != ERROR_BUFFER_OVERFLOW {
                // break the loop and check the return code again
                break;
            }
            // if the given memory size is too small to hold the adapter information,
            // the SizePointer returned will point to the required size of the buffer,
            // and we should continue.
            libc::free(buf as _);
```

Reply:
Actually, I would go even further and instead completely rewrite how `buf is handled by simplifying the `free`:

```rust
pub(crate) fn get_interface_address() -> Result<InterfaceAddressIterator, String> {
    unsafe {
        // https://learn.microsoft.com/en-us/windows/win32/api/iphlpapi/nf-iphlpapi-getadaptersaddresses#remarks
        // A 15k buffer is recommended
        let mut size: u32 = 15 * 1024;
        let mut buf = null_mut();

        // https://learn.microsoft.com/en-us/windows/win32/api/iphlpapi/nf-iphlpapi-getadaptersaddresses#examples
        // Try to retrieve adapter information up to 3 times
        for _ in 0..3 {
            buf = libc::realloc(buf, size as _) as PIP_ADAPTER_ADDRESSES;
            if buf.is_null() {
                // insufficient memory available
                // https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/malloc?view=msvc-170#return-value
                // malloc is not documented to set the last-error code
                return Err("failed to allocate memory for IP_ADAPTER_ADDRESSES".to_string());
            }

            let ret = GetAdaptersAddresses(
                AF_UNSPEC as u32,
                GAA_FLAG_SKIP_MULTICAST | GAA_FLAG_SKIP_ANYCAST | GAA_FLAG_SKIP_DNS_SERVER,
                null_mut(),
                buf,
                &mut size,
            );

            if ret == ERROR_SUCCESS {
                return Ok(InterfaceAddressIterator { buf, adapter: buf });
            }
            // We free the memory 
            if ret == ERROR_BUFFER_OVERFLOW {
                // if the given memory size is too small to hold the adapter information,
                // the SizePointer returned will point to the required size of the buffer,
                // and we should continue.
                continue
            }
            break;
        }

        // We free the memory since it won't be used anymore at this point.
        libc::free(buf as _);

        Err(format!("GetAdaptersAddresses() failed with code {ret}"))
    }
}
````

Reply:
never thought of `realloc`. I will try it

Reply:
```rust
buf = libc::realloc(buf, size as _) as PIP_ADAPTER_ADDRESSES;
```
realloc can [fail and leave `buf` unchanged](https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/realloc?view=msvc-170#return-value). It seems we still need to write code like

```rust
            let new_buf = libc::realloc(buf, size as _) as ...;
            if new_buf.is_null() {
                libc::free(buf); // feeding a null pointer is ok
                return Err(...);
            } else {
                buf = new_buf;
            }
```
`InterfaceAddressIterator` has already implemented `Drop`. Maybe I can use this type to enforce RAII

Reply:
We can wrap `realloc` into a function which will do it for us and return an `Err` in case it failed. That would look like:

```rust
let buf = realloc_wrapper(buf, size as _)?;

fn realloc_wrapper(buf: PIP_ADAPTER_ADDRESSES, new_size: usize) -> Result<PIP_ADAPTER_ADDRESSES, &'static str> {
    let new_buf = libc::realloc(buf, size as _) as ...;
    if new_buf.is_null() {
        libc::free(buf); // feeding a null pointer is ok
        Err(...)
    } else {
        Ok(new_buf)
    }
}
```